imap_processing/swe/l1b/swe_l1b_science.py

```diff
@@ -0,0 +1,509 @@
+                # Put counts rate in full cycle data array
+                full_cycle_data[esa_voltage_row_index][column_index] = counts_rate[step]
+                esa_step_number += 1
+
+            # reset column index for next quarter cycle
+            column_index = -1
+        # TODO: Apply in flight calibration to full cycle data
+
+    # NOTE: We may get more lookup table with different setup when we get real
+    # data. But for now, we are advice to continue with current setup and can
+    # add/change it when we get real data.
+
+    return full_cycle_data
+
+
+def find_cycle_starts(cycles: np.array):
+    """Find index of where new cycle started.
+
+    Brandon Stone helped developed this algorithm.
+
+    Parameters
+    ----------
+    cycles : np.array
+        Array that contains quarter cycle information.
+
+    Returns
+    -------
+    np.array
+        Array of indices of start cycle
+    """
+    if cycles.size < 4:
+        return np.array([], np.int64)
+
+    # calculate difference between consecutive cycles
+    diff = cycles[1:] - cycles[:-1]
+
+    # This uses sliding window to find index where cycle starts.
+    # This is what this below code line is doing:
+    # [1 0 0 1 0 0 0 0 0 1 0 0 1 0 0 0 0]      # Is cycle zero?
+    # [1 1 0 1 1 1 0 1 0 0 1 0 1 1 1 0 1]      # Next diff is one?
+    #   [1 0 1 1 1 0 1 0 0 1 0 1 1 1 0 1 0]    # Next diff is one?
+    #     [0 1 1 1 0 1 0 0 1 0 1 1 1 0 1 0 0]  # Next diff is one?
+    #
+    # [0 0 0 1 0 0 0 0 0 0 0 0 1 0 0 0 0]      # And all?
+    ione = diff == 1
+    valid = (cycles == 0)[:-3] & ione[:-2] & ione[1:-1] & ione[2:]
+    return np.where(valid)[0]
+
+
+def get_indices_of_full_cycles(quarter_cycle: np.array):
+    """Get indices of full cycles.
+
+    Parameters
+    ----------
+    quarter_cycle : np.array
+        Array that contains quarter cycles informations.
+
+    Returns
+    -------
+    np.array
+        1D array with indices of full cycle data.
+    """
+    indices_of_start = find_cycle_starts(quarter_cycle)
+    # indices_of_start[..., None] creates array of shape(n, 1).
+    #   Eg. [[3], [8]]
+    # np.arange(4)[None, ...] creates array of shape(1, 4)
+    #   Eg. [[0, 1, 2, 3]]
+    # then we add both of them together to get an array of shape(n, 4)
+    #   Eg. [[3, 4, 5, 6], [8, 9, 10, 11]]
+    full_cycles_indices = indices_of_start[..., None] + np.arange(4)[None, ...]
+    return full_cycles_indices.reshape(-1)
+
+
+def filter_full_cycle_data(full_cycle_data_indices: np.array, l1a_data: xr.Dataset):
+    """Filter metadata and science of packets that makes full cycles.
+
+    Parameters
+    ----------
+    full_cycle_data_indices : np.array
+        Array with indices of full cycles.
+    l1a_data : xr.dataset
+        L1A dataset
+
+    Returns
+    -------
+    xr.dataset
+        L1A dataset with filtered metadata.
+    """
+    for key, value in l1a_data.items():
+        l1a_data[key] = value.data[full_cycle_data_indices]
```

Comment:
I have some questions about that. Can I revisit this after thanksgiving break? I like to merge in this PR as soon as possible so that others can use common utils from this PR. I appreciate all the suggestions. I learned a lot in this PR!

Reply:
Yes, that is no problem to address it later.